### PR TITLE
Support additional protocols on top of P2P connection

### PR DESCRIPTION
**Pull request by karimodm**
# Description of change

This PR splits the P2P connection and neighbors managing component from the Gossip protocol.
While this set of changes it is not intended to alter any existing functionality, it paves the way to implement additional libp2p protocols on top of the existing P2P connection by introducing a top-level p2p Manager that:

* Handles neighbor connections.
* Allows to register protocol handlers.

Gossip is currently the only protocol registered.

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

All existing unit and integration tests pass.

### CODE REVIEW

File: packages/p2p/errors.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package p2p
+
+import "github.com/cockroachdb/errors"
+
+var (
+	// ErrNotRunning is returned when a neighbor is added to a stopped or not yet started gossip manager.
```
Comment (jonastheis) [inline packages/p2p/errors.go:6]: Please check all comments for gossip

File: packages/p2p/stream.go [hunk 2]
```diff
@@ -54,38 +54,38 @@ func (m *Manager) dialPeer(ctx context.Context, p *peer.Peer, opts []ConnectPeer
...
-		"addr", stream.Conn().RemoteMultiaddr(),
-	)
-	return ps, nil
+
+	return streams, err
 }
 
-func (m *Manager) acceptPeer(ctx context.Context, p *peer.Peer, opts []ConnectPeerOption) (*packetsStream, error) {
+func (m *Manager) acceptPeer(ctx context.Context, p *peer.Peer, opts []ConnectPeerOption) (map[protocol.ID]*PacketsStream, error) {
 	gossipEndpoint := p.Services().Get(service.GossipKey)
```
Comment (jonastheis) [inline packages/p2p/stream.go:83]: That is not gossip anymore no? Should be p2p

File: packages/p2p/stream.go [hunk 3]
```diff
@@ -54,38 +54,38 @@ func (m *Manager) dialPeer(ctx context.Context, p *peer.Peer, opts []ConnectPeer
...
-	if err != nil {
-		return nil, errors.Wrapf(err, "dial %s / %s failed", address, p.ID())
-	}
-	ps := newPacketsStream(stream)
-	if err := sendNegotiationBlock(ps); err != nil {
-		err = errors.Wrap(err, "failed to send negotiation block")
-		err = errors.CombineErrors(err, stream.Close())
-		return nil, err
+	streams := make(map[protocol.ID]*PacketsStream)
+	for protocolID, handlers := range m.registeredProtocols {
```
Comment (jonastheis) [inline packages/p2p/stream.go:66]: it's only a single handler no?

File: packages/p2p/stream.go [hunk 4]
```diff
@@ -54,38 +54,38 @@ func (m *Manager) dialPeer(ctx context.Context, p *peer.Peer, opts []ConnectPeer
...
-		"id", p.ID(),
-		"addr", stream.Conn().RemoteMultiaddr(),
-	)
-	return ps, nil
+
+	return streams, err
 }
 
-func (m *Manager) acceptPeer(ctx context.Context, p *peer.Peer, opts []ConnectPeerOption) (*packetsStream, error) {
+func (m *Manager) acceptPeer(ctx context.Context, p *peer.Peer, opts []ConnectPeerOption) (map[protocol.ID]*PacketsStream, error) {
```
Comment (jonastheis) [inline packages/p2p/stream.go:82]: How does `acceptPeer` work with multiple protocols? In `dialPeer` we're calling the `handlers.StreamEstablishFunc` multiple times. But here it looks like we're just doing it once. Are we calling `acceptPeer` multiple times then from the outside?

File: packages/gossip/manager.go [hunk 5]
```diff
@@ -4,69 +4,43 @@ import (
...
+	"github.com/iotaledger/goshimmer/packages/ratelimiter"
+	"github.com/iotaledger/goshimmer/packages/tangle"
 	"github.com/iotaledger/hive.go/generics/event"
 	"github.com/iotaledger/hive.go/identity"
 	"github.com/iotaledger/hive.go/logger"
-	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
 	libp2ppeer "github.com/libp2p/go-libp2p-core/peer"
 	"go.uber.org/atomic"
+	"google.golang.org/protobuf/proto"
```
Comment (jonastheis) [inline packages/gossip/manager.go:20]: imports not grouped/sorted properly

File: packages/gossip/manager.go [hunk 6]
```diff
@@ -137,73 +109,46 @@ func (m *Manager) Stop() {
...
-
-// GetNeighbor returns the neighbor by its id.
-func (m *Manager) GetNeighbor(id identity.ID) (*Neighbor, error) {
-	m.neighborsMutex.RLock()
-	defer m.neighborsMutex.RUnlock()
-	nbr, ok := m.neighbors[id]
-	if !ok {
-		return nil, ErrUnknownNeighbor
+	ps := p2p.NewPacketsStream(stream, gossipPacketFactory)
+	if err := sendNegotiationBlock(ps); err != nil {
```
Comment (jonastheis) [inline packages/gossip/manager.go:121]:
```suggestion
	if err := sendNegotiationMessage(ps); err != nil {
```

File: plugins/p2p/parameters.go [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package p2p
+
+import (
+	"github.com/iotaledger/hive.go/configuration"
+)
+
+// ParametersDefinition contains the definition of configuration parameters used by the gossip plugin.
```
Comment (jonastheis) [inline plugins/p2p/parameters.go:7]: fix comments

File: plugins/p2p/plugin.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/iotaledger/hive.go/autopeering/peer"
+	"github.com/iotaledger/hive.go/daemon"
+	"github.com/iotaledger/hive.go/generics/event"
+	"github.com/iotaledger/hive.go/node"
+
+	"github.com/iotaledger/goshimmer/packages/p2p"
+	"github.com/iotaledger/goshimmer/packages/shutdown"
+)
+
+// PluginName is the name of the gossip plugin.
```
Comment (jonastheis) [inline plugins/p2p/plugin.go:15]: fix comments

**Review by jonastheis — CHANGES_REQUESTED** [review]

File: packages/p2p/stream.go [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+	ErrTimeout = errors.New("accept timeout")
+	// ErrDuplicateAccept is returned when the server already registered an accept request for that peer ID.
+	ErrDuplicateAccept = errors.New("accept request for that peer already exists")
+	// ErrNoP2P means that the given peer does not support the p2p service.
+	ErrNoP2P = errors.New("peer does not have a p2p service")
+)
+
+func (m *Manager) dialPeer(ctx context.Context, p *peer.Peer, opts []ConnectPeerOption) (map[protocol.ID]*PacketsStream, error) {
+	m.registeredProtocolsMutex.RLock()
+	defer m.registeredProtocolsMutex.RUnlock()
```
[the remark this answers is not in the document]
Comment (karimodm) [reply packages/p2p/stream.go:43]: maps are references, so a getter won't prevent concurrent read/writes of the map. As this is a RLock, acquiring it later doesn't really make almost any difference: just in those cases where some error condition is met.

**Pull request merged**